### ISSUE TITLE
pyproject.toml based wheel builds

### ISSUE THREAD

**Issue by EpicWink** [google/jax#15256]
To support modern packaging standards (aka [PEP 518](https://peps.python.org/pep-0518/)) for wheel building, specify the build requirements and backend in [pyproject.toml](https://packaging.python.org/en/latest/glossary/#term-pyproject.toml).

I've tested following the instructions in the [setuptools documentation](https://setuptools.pypa.io/en/latest/userguide/quickstart.html#basic-use) (adding the `build-system` table), and installation from sdist works perfectly.

This will fix installation when there are no wheels provided and [setuptools](https://pypi.org/project/setuptools/) is not yet installed.

**Comment by jakevdp** [comment]
Thanks - just to be clear, you're talking about the `jax` package only, not the `jaxlib` package, yes? (The latter uses a bazel build system, and I don't think it's easily compatible with pyproject.toml)

**Comment by EpicWink** [comment]
@jakevdp yes, just `jax`

**Comment by jakevdp** [comment]
Thanks for the clarification - I think that may be possible, but will be somewhat painful. Our current setup process makes use of a lot of the imperative computation flexibility of `setup.py` files, for example in building the set of extra installation arguments: https://github.com/google/jax/blob/8c4fed64100924fea8a7e6d21223965acef3aaa1/setup.py#L128-L133

**Comment by EpicWink** [comment]
You can still use setup.py to declare package configuration and project metadata. My proposal is to add the the lines necessary to explicitly specifically setuptools as the build backend.

I think what you're thinking of is declaring metadata in pyproject.toml (in the `project` table), which is not what I'm suggesting. My proposal would have no changes to setup.py.

**Comment by eli-schwartz** [comment]
PEP 518 doesn't replace setup.py, it only defines how to say "build this by installing setuptools and running the setuptools script".

Also, the default for pep 518 if you don't add a pyproject.toml is to assume that the answer is setuptools so this is probably not actually necessary anyway...

**Comment by jakevdp** [comment]
OK, thanks. So what you have in mind is something like this? https://github.com/google/jax/pull/15274

**Comment by EpicWink** [comment]
> Also, the default for pep 518 if you don't add a pyproject.toml is to assume that the answer is setuptools so this is probably not actually necessary anyway...

Only if setuptools is installed, thus this feature request

**Comment by EpicWink** [comment]
> This sounds like your installation is not compliant with the part of PEP 518 that says:

PEPs are historical documents, not up-to-date specifications (with the exception of PEPs 1, 8, etc). The community is getting pretty close to not assuming setuptools is installed (which pip requires for the inferred pyproject.toml).

**Comment by eli-schwartz** [comment]
Pip will only assume that a project without a pyproject.toml file is a setuptools project, and consequently to seed an isolated environment with a brand new private copy of setuptools and fire off the setuptools build-backend API, if setuptools is installed outside of the isolated env?

That is an... interesting decision, and seems fairly arbitrary. It's not obvious to me why pip requires it... maybe "insists" would be a better word than "requires" as it lacks the implied idea of "needs and cannot otherwise implement"?

Maybe pip doesn't try to handle this case at all and activates legacy direct invocation of `python setup.py install`?

But that isn't true either, because testing with current versions of setuptools I observe the following behavior:
- if setuptools is not installed, then pip install will say "preparing metadata (pyproject.toml) ... done" even though I don't have any such file (I just deleted it in my own test project for this test)
- if I then pip install setuptools and then up-arrow twice and rerun `pip install .`, it prints "preparing metadata (setup.py) ... done" and then prints out an additional deprecation telling me that the legacy method is being used "because it does not have a pyproject.toml and the wheel package is not installed". In the same message it suggests that I can opt into a future behavior of pip today, run with --use-pep517, and it will act just like when setuptools isn't installed -- it will install both setuptools and wheel into an isolated env and build a wheel using that. 

> The community is getting pretty close to not assuming setuptools is installed 

Which is fantastic but also I am apparently missing something as I was under the impression that PEP 517/518 was never the problem there, and indeed I can't actually duplicate the issue of pip assuming setuptools is installed.

**Comment by eli-schwartz** [comment]
For the record, I still do think:
- explicit is better than implicit
- handling wheels even when setuptools is installed is good behavior
- using the modern build backend instead of the `setuptools.build_meta:__legacy__` one is good, because `__legacy__` is :( :(

It's just not clear to me who, exactly, cannot even build this at all right now.

**Comment by EpicWink** [comment]
The requirement (to get the error that caused me to create this issue) is there has to be a setup.py, no pyproject.toml, setuptools isn't installed, and only an sdist (and no wheel) available for the version. I don't think it matters whether `wheel` is installed or not.

Pip will assume it's a setuptools project when there is only a setup.py available, and attempt to run it directly (and it attempts to import the non-existent setuptools). I don't think pip will automatically install setuptools in this case.

**Comment by eli-schwartz** [comment]
> The requirement (to get the error that caused me to create this issue) is there has to be a setup.py, no pyproject.toml, setuptools isn't installed, and only an sdist (and no wheel) available for the version. I don't think it matters whether `wheel` is installed or not.

This combination of factors works for me when using pip 22.3.1.

```
rm pyproject.toml
ls # lists setup.py
python -m build --sdist

python -m venv /tmp/myvenv
. /tmp/myvenv/bin/activate

pip uninstall setuptools
pip list # only lists pip

pip install dist/*.tar.gz
```

produces:
```
Processing ./dist/myproject-1.2.3.tar.gz
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: myproject
  Building wheel for myproject (pyproject.toml) ... done
  Created wheel for myproject: filename=myproject-1.2.3-py3-none-any.whl size=915597 sha256=cfd94575c6cd086ea3da10e3e527de854e382710d82de6459070ae016b868c15
  Stored in directory: /home/eschwartz/.cache/pip/wheels/bd/81/0a/c26be8a25ee21746124f8f666b904fab82e4be890d98d8d9f0

...
```

It works for jax too:
```
$ /tmp/myvenv/bin/python -m pip list
Package Version
------- -------
pip     22.3.1

[notice] A new release of pip available: 22.3.1 -> 23.0.1
[notice] To update, run: python -m pip install --upgrade pip

$ /tmp/myvenv/bin/python -m pip install jax
Collecting jax
  Downloading jax-0.4.7.tar.gz (1.2 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 6.1 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting numpy>=1.21
  Downloading numpy-1.24.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.3/17.3 MB 10.1 MB/s eta 0:00:00
Collecting ml-dtypes>=0.0.3
  Downloading ml_dtypes-0.0.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (154 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 154.9/154.9 kB 1.7 MB/s eta 0:00:00
Collecting opt-einsum
  Downloading opt_einsum-3.3.0-py3-none-any.whl (65 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.5/65.5 kB 811.4 kB/s eta 0:00:00
Collecting scipy>=1.7
  Downloading scipy-1.10.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (34.4 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 34.4/34.4 MB 8.0 MB/s eta 0:00:00
Building wheels for collected packages: jax
  Building wheel for jax (pyproject.toml) ... done
  Created wheel for jax: filename=jax-0.4.7-py3-none-any.whl size=1438271 sha256=4d4836863608dce6f825c0d377561caf70e46aa6ba73d3da3624148e7ef675f2
  Stored in directory: /home/eschwartz/.cache/pip/wheels/8e/ee/24/ebc3556952db522adf4946bfff25ae4cc02cf40d6992cf9d42
Successfully built jax
Installing collected packages: numpy, scipy, opt-einsum, ml-dtypes, jax
Successfully installed jax-0.4.7 ml-dtypes-0.0.4 numpy-1.24.2 opt-einsum-3.3.0 scipy-1.10.1

[notice] A new release of pip available: 22.3.1 -> 23.0.1
[notice] To update, run: python -m pip install --upgrade pip
```

It is claiming to build jax via pyproject.toml, today.
